### ISSUE TITLE
Help to turn this into a Dropwizard module

### ISSUE THREAD

**Issue by grange74** [ahus1/keycloak-dropwizard-integration#1]
First of all great work in putting this code together and the excellent documentation.

You mention that this will eventually become a Dropwizard Module. I'm looking to use Keycloak with Dropwizard so i'd like to see if i can help you make this happen.

I've tried out your keycloak-dropwizard-jaxrs code by [forking it](https://github.com/grange74/keycloak-dropwizard-integration/tree/GRADLE-MODULE/keycloak-dropwizard-jaxrs/src/main/java/de/ahus1/lottery/adapter/dropwizard) and removing all non-module related code and then using it as a dependency in a [very simple Dropwizard project](https://github.com/grange74/dropwizard-examples/tree/master/keycloak/src/main/java/com/example/keycloak). The code works perfectly as expected.

Btw do you plan on creating a new repo for the module or just split up the existing keycloak-dropwizard-jaxrs folder into 2 (a module and a sample project)?

Anyway let me know if/how i can help.

**Comment by ahus1** [comment]
Thank you for trying out the examples. There are multiple options and I was wondering what other people or myself should use. I thought that the keycloak-dropwizard-jaxrs solution was still too complex, and I was waiting for a first feedback. 

As the keycloak-dropwizard-jaxrs module works for you it's now a good time to properly package it. I have split up keycloak-dropwizard-jaxrs and keycloak-dropwizard-jaxrs-example into separate modules.

You'll also notice that the code changed to a real bundle. Here the excerpt from the example: 

        bootstrap.addBundle(new KeycloakBundle<LotteryConfiguration>() {
            @Override
            protected KeycloakAuthFactory createAuthFactory(LotteryConfiguration configuration) {
                return new KeycloakAuthFactory(getKeycloakConfiguration(configuration), "dropwizard",
                        new KeycloakAuthenticator(), Authentication.class);
            }

            @Override
            protected AdapterConfig getKeycloakConfiguration(LotteryConfiguration configuration) {
                return configuration.getKeycloakConfiguration();
            }
        });

The idea behind it: 

   * You can use your own `Authentication` class (and due to this `KeycloakAuthFactory`) in your application so you can make `Authentication` work for you (for example any custom or complex role checks you need for your domain).

Please let me know what you think. Either comment here, or open a new ticket for a new topic. Feel free to comment on naming of artefacts, packages and classes. Once we agree it is ready I'll roll a first release.

**Comment by grange74** [comment]
Just tried out your changes and they work well.

Here is some minor feedback:
* Make class AbstractAuthentication abstract
* Ideally people should require to import/use Keycloak specific classes directly. Atleast to start with. I know that this isn't easy because it would mean writing wrappers for classes like KeycloakSecurityContext, IDToken and AdapterConfig or just extending them.
* It is great that you've thought about Extensibility with the Abstract classes. I would recommend also including the Default implementations in the module so someone can get started with just adding the configuration to their yml file, adding a small amount of code to their Configuration and Application classes and then just securing the appropriate resources with the Authenticator param.
* for me the Authentication class could be called User https://dropwizard.github.io/dropwizard/manual/auth.html#protecting-resources
* with the package/module name i would exclude "jaxrs" (e.g. just "dropwizard-keycloak") people generally choose Dropwizard to create RESTful Services

**Comment by ahus1** [comment]
Hello Nicolas,

thank you for your feedback. I have applied it to the example project as well. All tests are green. 

There is now a minimal default implementation of User and KeycloakAuthenticator.

I agree that applications should not depend on Keycloak classes in their REST endpoints. The key to this (as I see it) would be for each application to implement its own implementation of User. This class would provide a business level API, maybe starting with a type safe domain-specific role checker (while the standard User just receives a String as role name). Only the User class would be allowed to talk the IDToken/AccessToken.

If your examples need some more methods in the default User implementation please go ahead to to send a PR. 

Please keep the discussion about structure and naming active in this issue, or close it when you are happy. I will then look into releasing it. 

Please note that I have also updated the dependency to Keycloak 1.5 as it allowed me to remove a workaround for a Keycloak issue. I still have to migrate the keycloak-realm.json file to Keycloak 1.5.

Best regards,
Alexander

**Comment by grange74** [comment]
Hi Alexander,

I just tried out your updates and they look great. It is now super easy to get started while still being extensible.

I will happily submit PRs for the User class in the near future when i start using the module for a real project.

Regarding keeping this issue open, the only remaining structure and naming suggestion i have is to create a KeycloakConfiguration class to wrap the AdapterConfig class. 
I did a quick prototype of it on this branch: https://github.com/grange74/keycloak-dropwizard-integration/commit/92f83dfea56f5aba44d66f0ee90c0a35ff2040f6 and the sample code would look like this https://github.com/grange74/dropwizard-examples/commit/13bf3fb6354ea964a212c28899b21f29af3ed239.

Other than this i think it is good enough to release the first version.

You should consider submitting a PR for putting it on the 3rd party modules page (i'm happy to do it as well). 
http://modules.dropwizard.io/thirdparty/

I have one other improvement but it is unrelated to structure and naming so i will create another issue.

Btw it is a pleasure to work with someone so positive and passionate about what they are doing.

Cheers,
Nic

**Comment by ahus1** [comment]
Hello Nic,

I changed the Config as you suggested. I would have been opposed to a deep copy of the configuration as I would have feared to miss a property either now or when the Keycloak interface changes in the future. 

You chose inheritance to avoid this. But with this approach the structure of the configuration file might change with every keycloak version change. Let's see ... time will tell. 

Thank you for all the good input and discussion!

Cheers,
Alexander

**Comment by ahus1** [comment]
I forgot: adding it to the modules was under way already. https://github.com/dropwizard/modules.dropwizard.io/pull/20 - this pull request has been merged already.
